src/io/fastq.rs

```diff
@@ -208,45 +208,56 @@ where
     /// assert_eq!(record.qual().to_vec(), b"IIII");
     /// ```
     fn read(&mut self, record: &mut Record) -> io::Result<()> {
+        let incomplete_msg = "Incomplete record. Each FastQ record has to consist \
+                     of 4 lines: header, sequence, separator and \
+                     qualities.";
+        
         record.clear();
         self.line_buffer.clear();
-
-        self.reader.read_line(&mut self.line_buffer)?;
-
-        if !self.line_buffer.is_empty() {
+        if self.reader.read_line(&mut self.line_buffer)? > 0 { // reader successfully read bytes
             if !self.line_buffer.starts_with('@') {
                 return Err(io::Error::new(
                     io::ErrorKind::Other,
                     "Expected @ at record start.",
                 ));
             }
+            let read_line = |reader: &mut io::BufReader<R>, line_buffer: &mut String| {
+                if reader.read_line(line_buffer)? == 0 {
+                    // eof reached
+                    return Err(io::Error::new(
+                        io::ErrorKind::Other,
+                        incomplete_msg,
+                    ));
```

Comment:
It'd be nice if we could have a custom `Error` for this. I'm slightly out of the loop when it comes to error handling in rust-bio. I think it's mostly `snafu` but in rust-htslib we switched to `thiserror`.